### ISSUE TITLE
Died on user notice of user warning - reading property textContent for AntiSpywareProfile

### ISSUE THREAD

**Issue by bethatasitmay** [swaschkut/pan-os-php#19]
## Describe the bug

I updated my pan-os-php install (archive current install folder & re-retrieve current release - see below) from a download of whatever was current on 12 April 2024 to the current release of 18 June 2024 (both 2.1.24 versions but different builds) and now I get this error:

** ERROR ** * Died on user notice or warning!! Error: Attempt to read property "textContent" on false on C:\tools\pan\pan-os-php\lib\object-classes\AntiSpywareProfile.php:304

## Expected behavior

Whenever we install the latest build, we run this command to confirm the install is working (in other words, this has worked for possibly as long as pan-os-php has existed):

> php -r "require_once 'C:/tools/pan/pan-os-php/utils/pan-os-php.php';" type=rule in=api://panorama.domain.com location=firewall-1 ruletype=security actions=display 'filter=(name contains rulename-)'

"rulename-" is sanitized - there is really a unique rulename prefix that is contained in a number of permanent rules so we always get results for it.

Update procedure:

* Rename folder containing pano-os-php (e.g., rename to pan-os-php-2.1.24-A)
* In PowerShell, run: git clone https://github.com/swaschkut/pan-os-php.git

After updating, we run the command above.

Normally, the result is something like this (this is a different command as I don't have a saved copy of a working attempt of the command above and also this output below is from the April download that I ran just prior to updating to the current release).  The command below is one we use to identify the next available number in incremented rules (e.g., rule1, rule2, and rule3 are used so we test for rule4 as a guess to see if the display command returns a null result (0 filtered/0 objects over <number> available).

> php -r "require_once 'C:/tools/pan/pan-os-php/utils/pan-os-php.php';" type=rule in=api://panorama.domain.com location=any ruletype=security actions=display 'filter=(name contains pre-rule999)'

***********************************************
*********** pan-os-php.php type=rule UTILITY **************

 - PAN-OS-PHP version: 2.1.24 [WIN] [8.3.8]

####################################
For PAN-OS version < 9.0 please use additional argument 'shadow-apikeynohidden' in your script command

####################################

 - Downloading config from API...
 - Detected platform type is 'panorama'
 - No 'template' provided so using default ='any'
 - filter after sanitization : (name contains pre-rule999)
 - Loading configuration through PAN-OS-PHP library...
   (1.30 seconds, 60.82 mb memory)
 - PAN-OS-PHP APP-ID version: 8833-8682

 - PAN-OS APP-ID version:
 - PAN-OS AV version:
 - PAN-OS WF version:
 - PAN-OS THREAT version:
 - PAN-OS version: 101
 - PAN-OS Device timezone: US/Pacific is used. actual time: 2024/07/02 14:48:40

* processing ruleset 'PanoramaConf: / RuleStore:Security' that holds 461 rules
* objects processed in DG/Vsys '' : 0 filtered over 461 available

* processing ruleset 'PanoramaConf: / DeviceGroup:firewall-2 / RuleStore:Security' that holds 1116 rules
* objects processed in DG/Vsys 'firewall-2' : 0 filtered over 1116 available

* processing ruleset 'PanoramaConf: / DeviceGroup:firewall-1 / RuleStore:Security' that holds 1776 rules
* objects processed in DG/Vsys 'firewall-1' : 0 filtered over 1776 available

 **** PROCESSED 0 objects over 3353 available ****

argument 'actions' was used with 'display'

************* END OF SCRIPT pan-os-php.php type=rule ************

However, I don't remember those blank PAN-OS lines being unpopulated in previous runs (but I can't be sure) which is what prompted me to update to the current release (I usually update within days to months unless I run into a problem).

## Current behavior

This was just run on the current release and has always worked in previous releases since we started using PowerShell instead of Docker which we're not able to use due 

> php -r "require_once 'C:/tools/pan/pan-os-php/utils/pan-os-php.php';" type=rule in=api://panorama.domain.com location=firewall-1 ruletype=security actions=display 'filter=(name contains rulename-)'

***********************************************
*********** pan-os-php.php type=rule UTILITY **************

 - PAN-OS-PHP version: 2.1.24 [WIN] [8.3.8]
 - Downloading config from API...
 - Detected platform type is 'panorama'
 - No 'template' provided so using default ='any'
 - filter after sanitization : (name contains rulename-)
 - Loading configuration through PAN-OS-PHP library...

* ** ERROR ** * Died on user notice or warning!! Error: Attempt to read property "textContent" on false on C:\tools\pan\pan-os-php\lib\object-classes\AntiSpywareProfile.php:304

 *** Backtrace ***
0 ****
       backtrace_print()
       ::C:\tools\pan\pan-os-php\lib\pan_php_framework.php line 617
1 ****
       derr()
       ::C:\tools\pan\pan-os-php\lib\pan_php_framework.php line 117
2 ****
       myErrorHandler()
       ::C:\tools\pan\pan-os-php\lib\object-classes\AntiSpywareProfile.php line 304
3 ****
   PanoramaConf: / SecurityProfileStore:AntiSpyware / AntiSpywareProfile:AS-MHC-Drop-NoCap
       AntiSpywareProfile::load_from_domxml() @
           C:\tools\pan\pan-os-php\lib\object-classes\SecurityProfileStore.php line 148
4 ****
   PanoramaConf: / SecurityProfileStore:AntiSpyware
       SecurityProfileStore::load_from_domxml() @
           C:\tools\pan\pan-os-php\lib\device-and-system-classes\PanoramaConf.php line 613
5 ****
   PanoramaConf:
       PanoramaConf::load_from_domxml() @
           C:\tools\pan\pan-os-php\utils\lib\UTIL.php line 1298
6 ****
       UTIL::load_config() @
           C:\tools\pan\pan-os-php\utils\lib\UTIL.php line 333
7 ****
       UTIL::utilActionFilter() @
           C:\tools\pan\pan-os-php\utils\lib\RULEUTIL.php line 31
8 ****
       RULEUTIL::utilStart() @
           C:\tools\pan\pan-os-php\utils\lib\UTIL.php line 215
9 ****
       UTIL::__construct() @
           C:\tools\pan\pan-os-php\lib\misc-classes\PH.php line 1051
10 ****
       callPANOSPHP()
       ::C:\tools\pan\pan-os-php\utils\pan-os-php.php line 118
11 ****
       require_once()
       ::Command line code line 1

If I restore the previous build I was using (installed in April) by renaming pan-os-php to pan-os-php-2.1.24-B and renaming pan-os-php-2.1.24-A back to pan-os-php, commands are working again including my test command noted immediately above.

Also, rolling back to the previous install shows the PAN-OS lines being populated.

Earlier today:

- PAN-OS-PHP APP-ID version: 8833-8682

 - PAN-OS APP-ID version:
 - PAN-OS AV version:
 - PAN-OS WF version:
 - PAN-OS THREAT version:
 - PAN-OS version: 101
 - PAN-OS Device timezone: US/Pacific is used. actual time: 2024/07/02 14:48:40

Then I updated to the June build and then I rolled back to my April download and re-ran our install test command:

 - PAN-OS-PHP APP-ID version: 8833-8682

 - PAN-OS APP-ID version: 8866-8815
 - PAN-OS AV version: 4868-5386
 - PAN-OS WF version: 515890-518890
 - PAN-OS THREAT version: 8866-8815
 - PAN-OS version: 101
 - PAN-OS Device timezone: US/Pacific is used. actual time: 2024/07/03 11:31:13

## Context

The failed run is just a test to see if a new install is working, however I re-ran the 2nd command ("pre-rule999") along with a previously-known-working command and they return the same error messages.

So, basically, I can no longer use the tool.

## Your Environment

- Version used: 2.1.24, release 18 June 2024 (not working); 2.1.24, current build that was available on 12 April 2024
- Environment name and version: PowerShell v5
- Operating System and version: Windows 10

**Comment by swaschkut** [comment]
thanks for letting me know,
I will check and bring in additional validation

**Comment by swaschkut** [comment]
please validate if new developer code is working fine:

https://github.com/swaschkut/pan-os-php/blob/develop_2-1-X

**Comment by bethatasitmay** [comment]
Yes, developer code works (including the PAN-OS lines being populated)!  Thanks again (and again and again :-) )!

**Comment by bethatasitmay** [comment]
By the way, the developer code also fixed another issue I was having and was about to report.

I was trying to use action group-tag-set and on the 2.1.24 release (the one I installed in April) I was getting this...

> php -r "require_once 'C:/tools/pan/pan-os-php/utils/pan-os-php.php';" type=rule in=api://panorama.domain.com location=firewall-2 actions=group-tag-Set:TAG1 'filter=(name contains rulename-number)'

***********************************************
*********** pan-os-php.php type=rule UTILITY **************

 - PAN-OS-PHP version: 2.1.24 [WIN] [8.3.8]
 - Downloading config from API...
 - Detected platform type is 'panorama'
 - No 'template' provided so using default ='any'
 - No 'ruleType' specified, using 'security' by default
 - filter after sanitization : (name contains rulename-number)
 - Loading configuration through PAN-OS-PHP library...
   (1.08 seconds, 60.82 mb memory)
 - PAN-OS-PHP APP-ID version: 8833-8682

 - PAN-OS APP-ID version: 8867-8824
 - PAN-OS AV version: 4868-5386
 - PAN-OS WF version: 515890-518890
 - PAN-OS THREAT version: 8867-8824
 - PAN-OS version: 101
 - PAN-OS Device timezone: US/Pacific is used. actual time: 2024/07/03 19:10:42

* processing ruleset 'PanoramaConf: / DeviceGroup:firewall-2 / RuleStore:Security' that holds 1133 rules
   - object 'rulename-04'  passing through Action='group-tag-Set' Args: Group-Tag=TAG1,
PHP Fatal error:  Uncaught Error: Call to undefined method GroupTagRuleContainer::removeReference() in C:\tools\pan\pan-os-php\lib\rule-classes\RuleWithGroupTag.php:119
Stack trace:
#0 C:\tools\pan\pan-os-php\lib\rule-classes\RuleWithGroupTag.php(140): Rule->setGroupTag()
#1 C:\tools\pan\pan-os-php\utils\common\actions-rule.php(5461): Rule->API_setGroupTag()
#2 C:\tools\pan\pan-os-php\utils\common\CallContext.php(116): {closure}()
#3 C:\tools\pan\pan-os-php\utils\lib\RULEUTIL.php(493): CallContext->executeAction()
#4 C:\tools\pan\pan-os-php\utils\lib\RULEUTIL.php(37): RULEUTIL->time_to_process_objects()
#5 C:\tools\pan\pan-os-php\utils\lib\UTIL.php(214): RULEUTIL->utilStart()
#6 C:\tools\pan\pan-os-php\lib\misc-classes\PH.php(1050): UTIL->__construct()
#7 C:\tools\pan\pan-os-php\utils\pan-os-php.php(118): PH::callPANOSPHP()
#8 Command line code(1): require_once('...')
#9 {main}
  thrown in C:\tools\pan\pan-os-php\lib\rule-classes\RuleWithGroupTag.php on line 119

**Comment by swaschkut** [comment]
fixed with version 2.1.25
#21
